Uploaded picture corrupt (part is grey)

I uploaded a picture of a building, but unfortunately the bottom of the uploaded picture is grey: https://commons.wikimedia.org/wiki/File:%E6%9D%B1%E4%BA%AC%E5%A1%97%E6%96%99%E4%BC%9A%E9%A4%A8.jpg

I suspect that only the beginning of the file is uploaded, and then the upload gets considered as finished even though it is not. Depictions/categories/etc are set correctly.

The same actually happened twice last week too:

![tempFileForShare_20200817-233115](https://user-images.githubusercontent.com/99590/90407908-08050980-e0e2-11ea-8cb2-ec49d0acb192.jpg)

Using the Commons website I overwrote these two but the original upload can be seen in their history pages: https://commons.wikimedia.org/wiki/File:Restaurant_%E6%B3%89%E8%8F%AF%E8%8D%98_2.jpg https://commons.wikimedia.org/wiki/File:Phasme_%C3%A0_Tokyo.jpg

App version: 2.12.3-debug-temp-filteere~796f1f07d (very close to current master, only leaderboard addition that does not change anything to the upload part)

Below is the log for this corrupted upload: https://commons.wikimedia.org/wiki/File:Strange_eggplant.jpg
```
[... many identical SocketTimeoutExceptions ...]
2020-08-31 21:39:50,421     [ERROR] [file-logging-thread-1] [UploadClient] : Error occurred in uploading chunk
java.net.SocketTimeoutException: timeout
	at okhttp3.internal.http2.Http2Stream$StreamTimeout.newTimeoutException(Http2Stream.kt:677)
	at okhttp3.internal.http2.Http2Stream$StreamTimeout.exitAndThrowIfTimedOut(Http2Stream.kt:686)
	at okhttp3.internal.http2.Http2Stream$FramingSink.emitFrame(Http2Stream.kt:557)
	at okhttp3.internal.http2.Http2Stream$FramingSink.write(Http2Stream.kt:535)
	at okio.ForwardingSink.write(ForwardingSink.kt:29)
	at okhttp3.internal.connection.Exchange$RequestBodySink.write(Exchange.kt:218)
	at okio.RealBufferedSink.emitCompleteSegments(RealBufferedSink.kt:235)
	at okio.RealBufferedSink.write(RealBufferedSink.kt:145)
	at okio.ForwardingSink.write(ForwardingSink.kt:29)
	at fr.free.nrw.commons.upload.CountingRequestBody$CountingSink.write(CountingRequestBody.kt:49)
	at okio.RealBufferedSink.emitCompleteSegments(RealBufferedSink.kt:235)
	at okio.RealBufferedSink.writeAll(RealBufferedSink.kt:184)
	at okhttp3.RequestBody$Companion$asRequestBody$1.writeTo(RequestBody.kt:167)
	at fr.free.nrw.commons.upload.CountingRequestBody.writeTo(CountingRequestBody.kt:40)
	at okhttp3.MultipartBody.writeOrCountBytes(MultipartBody.kt:157)
	at okhttp3.MultipartBody.writeTo(MultipartBody.kt:93)
	at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.kt:59)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.kt:34)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.kt:95)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.kt:83)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.kt:76)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at fr.free.nrw.commons.OkHttpConnectionFactory$CommonHeaderRequestInterceptor.intercept(OkHttpConnectionFactory.java:63)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at fr.free.nrw.commons.OkHttpConnectionFactory$UnsuccessfulResponseInterceptor.intercept(OkHttpConnectionFactory.java:76)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.logging.HttpLoggingInterceptor.intercept(HttpLoggingInterceptor.kt:219)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:201)
	at okhttp3.internal.connection.RealCall.execute(RealCall.kt:154)
	at retrofit2.OkHttpCall.execute(OkHttpCall.java:190)
	at retrofit2.adapter.rxjava2.CallExecuteObservable.subscribeActual(CallExecuteObservable.java:42)
	at io.reactivex.Observable.subscribe(Observable.java:12267)
	at retrofit2.adapter.rxjava2.BodyObservable.subscribeActual(BodyObservable.java:34)
	at io.reactivex.Observable.subscribe(Observable.java:12267)
	at io.reactivex.internal.operators.observable.ObservableMap.subscribeActual(ObservableMap.java:32)
	at io.reactivex.Observable.subscribe(Observable.java:12267)
	at io.reactivex.Observable.subscribe(Observable.java:12253)
	at io.reactivex.Observable.subscribe(Observable.java:12181)
	at fr.free.nrw.commons.upload.UploadClient.lambda$uploadFileToStash$2$UploadClient(UploadClient.java:106)
	at fr.free.nrw.commons.upload.-$$Lambda$UploadClient$o9957uEjSG2yFmKju6TjCB8UUak.accept(Unknown Source:19)
	at io.reactivex.internal.observers.LambdaObserver.onNext(LambdaObserver.java:63)
	at io.reactivex.internal.operators.observable.ObservableFromIterable$FromIterableDisposable.run(ObservableFromIterable.java:98)
	at io.reactivex.internal.operators.observable.ObservableFromIterable.subscribeActual(ObservableFromIterable.java:58)
	at io.reactivex.Observable.subscribe(Observable.java:12267)
	at io.reactivex.Observable.subscribe(Observable.java:12253)
	at io.reactivex.Observable.subscribe(Observable.java:12155)
	at io.reactivex.Observable.forEach(Observable.java:9105)
	at fr.free.nrw.commons.upload.UploadClient.uploadFileToStash(UploadClient.java:83)
	at fr.free.nrw.commons.upload.UploadService.lambda$uploadContribution$2$UploadService(UploadService.java:269)
	at fr.free.nrw.commons.upload.-$$Lambda$UploadService$Mykz4B1wymjq4LCG_9JtKi5_-pY.apply(Unknown Source:8)
	at io.reactivex.internal.operators.observable.ObservableScalarXMap.tryScalarXMapSubscribe(ObservableScalarXMap.java:70)
	at io.reactivex.internal.operators.observable.ObservableFlatMap.subscribeActual(ObservableFlatMap.java:51)
	at io.reactivex.Observable.subscribe(Observable.java:12267)
	at io.reactivex.internal.operators.observable.ObservableSubscribeOn$SubscribeTask.run(ObservableSubscribeOn.java:96)
	at io.reactivex.Scheduler$DisposeTask.run(Scheduler.java:578)
	at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66)
	at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.lang.Thread.run(Thread.java:764)

2020-08-31 21:40:01,077     [ERROR] [file-logging-thread-1] [ContributionBoundaryCallback$fetchContributions] : Failed to fetch contributions: timeout
2020-08-31 21:40:01,564     [ERROR] [file-logging-thread-1] [ContributionBoundaryCallback$fetchContributions] : Failed to fetch contributions: timeout
2020-08-31 21:40:11,061     [ERROR] [file-logging-thread-1] [ContributionBoundaryCallback$fetchContributions] : Failed to fetch contributions: timeout
2020-08-31 21:40:11,935     [DEBUG] [file-logging-thread-1] [LocationServiceManager] : gps's status changed to 1
2020-08-31 21:40:20,638     [ERROR] [file-logging-thread-1] [UploadClient] : Error occurred in uploading chunk
java.net.SocketTimeoutException: timeout
	at okhttp3.internal.http2.Http2Stream$StreamTimeout.newTimeoutException(Http2Stream.kt:677)
	at okhttp3.internal.http2.Http2Stream$StreamTimeout.exitAndThrowIfTimedOut(Http2Stream.kt:686)
	at okhttp3.internal.http2.Http2Stream$FramingSink.emitFrame(Http2Stream.kt:557)
	at okhttp3.internal.http2.Http2Stream$FramingSink.write(Http2Stream.kt:535)
	at okio.ForwardingSink.write(ForwardingSink.kt:29)
	at okhttp3.internal.connection.Exchange$RequestBodySink.write(Exchange.kt:218)
	at okio.RealBufferedSink.emitCompleteSegments(RealBufferedSink.kt:235)
	at okio.RealBufferedSink.write(RealBufferedSink.kt:145)
	at okio.ForwardingSink.write(ForwardingSink.kt:29)
	at fr.free.nrw.commons.upload.CountingRequestBody$CountingSink.write(CountingRequestBody.kt:49)
	at okio.RealBufferedSink.emitCompleteSegments(RealBufferedSink.kt:235)
	at okio.RealBufferedSink.writeAll(RealBufferedSink.kt:184)
	at okhttp3.RequestBody$Companion$asRequestBody$1.writeTo(RequestBody.kt:167)
	at fr.free.nrw.commons.upload.CountingRequestBody.writeTo(CountingRequestBody.kt:40)
	at okhttp3.MultipartBody.writeOrCountBytes(MultipartBody.kt:157)
	at okhttp3.MultipartBody.writeTo(MultipartBody.kt:93)
	at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.kt:59)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.kt:34)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.kt:95)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.kt:83)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.kt:76)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at fr.free.nrw.commons.OkHttpConnectionFactory$CommonHeaderRequestInterceptor.intercept(OkHttpConnectionFactory.java:63)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at fr.free.nrw.commons.OkHttpConnectionFactory$UnsuccessfulResponseInterceptor.intercept(OkHttpConnectionFactory.java:76)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.logging.HttpLoggingInterceptor.intercept(HttpLoggingInterceptor.kt:219)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:201)
	at okhttp3.internal.connection.RealCall.execute(RealCall.kt:154)
	at retrofit2.OkHttpCall.execute(OkHttpCall.java:190)
	at retrofit2.adapter.rxjava2.CallExecuteObservable.subscribeActual(CallExecuteObservable.java:42)
	at io.reactivex.Observable.subscribe(Observable.java:12267)
	at retrofit2.adapter.rxjava2.BodyObservable.subscribeActual(BodyObservable.java:34)
	at io.reactivex.Observable.subscribe(Observable.java:12267)
	at io.reactivex.internal.operators.observable.ObservableMap.subscribeActual(ObservableMap.java:32)
	at io.reactivex.Observable.subscribe(Observable.java:12267)
	at io.reactivex.Observable.subscribe(Observable.java:12253)
	at io.reactivex.Observable.subscribe(Observable.java:12181)
	at fr.free.nrw.commons.upload.UploadClient.lambda$uploadFileToStash$2$UploadClient(UploadClient.java:106)
	at fr.free.nrw.commons.upload.-$$Lambda$UploadClient$o9957uEjSG2yFmKju6TjCB8UUak.accept(Unknown Source:19)
	at io.reactivex.internal.observers.LambdaObserver.onNext(LambdaObserver.java:63)
	at io.reactivex.internal.operators.observable.ObservableFromIterable$FromIterableDisposable.run(ObservableFromIterable.java:98)
	at io.reactivex.internal.operators.observable.ObservableFromIterable.subscribeActual(ObservableFromIterable.java:58)
	at io.reactivex.Observable.subscribe(Observable.java:12267)
	at io.reactivex.Observable.subscribe(Observable.java:12253)
	at io.reactivex.Observable.subscribe(Observable.java:12155)
	at io.reactivex.Observable.forEach(Observable.java:9105)
	at fr.free.nrw.commons.upload.UploadClient.uploadFileToStash(UploadClient.java:83)
	at fr.free.nrw.commons.upload.UploadService.lambda$uploadContribution$2$UploadService(UploadService.java:269)
	at fr.free.nrw.commons.upload.-$$Lambda$UploadService$Mykz4B1wymjq4LCG_9JtKi5_-pY.apply(Unknown Source:8)
	at io.reactivex.internal.operators.observable.ObservableScalarXMap.tryScalarXMapSubscribe(ObservableScalarXMap.java:70)
	at io.reactivex.internal.operators.observable.ObservableFlatMap.subscribeActual(ObservableFlatMap.java:51)
	at io.reactivex.Observable.subscribe(Observable.java:12267)
	at io.reactivex.internal.operators.observable.ObservableSubscribeOn$SubscribeTask.run(ObservableSubscribeOn.java:96)
	at io.reactivex.Scheduler$DisposeTask.run(Scheduler.java:578)
	at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66)
	at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.lang.Thread.run(Thread.java:764)

2020-08-31 21:40:27,742     [DEBUG] [file-logging-thread-1] [LocationServiceManager] : gps's status changed to 1
2020-08-31 21:40:31,244     [ERROR] [file-logging-thread-1] [ContributionBoundaryCallback$fetchContributions] : Failed to fetch contributions: timeout
2020-08-31 21:40:31,471     [ERROR] [file-logging-thread-1] [ContributionBoundaryCallback$fetchContributions] : Failed to fetch contributions: timeout
2020-08-31 21:40:31,491     [ERROR] [file-logging-thread-1] [ContributionBoundaryCallback$fetchContributions] : Failed to fetch contributions: timeout
2020-08-31 21:40:41,003     [ERROR] [file-logging-thread-1] [ContributionBoundaryCallback$fetchContributions] : Failed to fetch contributions: timeout
2020-08-31 21:40:47,924     [DEBUG] [file-logging-thread-1] [LocationServiceManager] : gps's status changed to 1
2020-08-31 21:41:09,195     [DEBUG] [file-logging-thread-1] [LocationServiceManager] : gps's status changed to 1
2020-08-31 21:41:31,365     [DEBUG] [file-logging-thread-1] [LocationServiceManager] : gps's status changed to 1
2020-08-31 21:41:59,428     [DEBUG] [file-logging-thread-1] [LocationServiceManager] : gps's status changed to 1
2020-08-31 21:42:27,676     [DEBUG] [file-logging-thread-1] [LocationServiceManager] : on location changed
2020-08-31 21:42:27,685     [DEBUG] [file-logging-thread-1] [LocationServiceManager] : gps's status changed to 1
2020-08-31 21:42:50,894     [DEBUG] [file-logging-thread-1] [UploadService] : Stash upload response 1 is StashUploadResult(state=SUCCESS, fileKey=17qbh2m96qi0.6nexud.27103.jpg)
2020-08-31 21:42:50,904     [DEBUG] [file-logging-thread-1] [UploadService] : making sure of uniqueness of name: strange eggplant.jpg
2020-08-31 21:42:55,402     [DEBUG] [file-logging-thread-1] [UploadService] : Stash upload response 2 is UploadResult(result=Success, filekey=null, offset=0, filename=Strange_eggplant.jpg)
2020-08-31 21:42:55,878     [DEBUG] [file-logging-thread-1] [WikidataEditService] : EntityId for image was received successfully: 93707744
2020-08-31 21:42:56,461     [DEBUG] [file-logging-thread-1] [LocationServiceManager] : gps's status changed to 1
2020-08-31 21:42:57,327     [DEBUG] [file-logging-thread-1] [WikidataEditService] : DEPICTS property was set successfully for 93707744
2020-08-31 21:42:57,329     [DEBUG] [file-logging-thread-1] [WikidataEditService] : edit response: true
2020-08-31 21:42:58,832     [DEBUG] [file-logging-thread-1] [WikidataEditService] : Caption successfully set, revision id = org.wikipedia.dataclient.mwapi.MwPostResponse@27c5a37[success=1,errors=<null>]
2020-08-31 21:42:58,836     [DEBUG] [file-logging-thread-1] [WikidataEditService] : edit response: true
```

--- a/app/src/main/java/fr/free/nrw/commons/OkHttpConnectionFactory.java
+++ b/app/src/main/java/fr/free/nrw/commons/OkHttpConnectionFactory.java
@@ -6,6 +6,7 @@ import java.io.IOException;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
+import java.util.concurrent.TimeUnit;
 import okhttp3.Cache;
 import okhttp3.Interceptor;
 import okhttp3.OkHttpClient;
@@ -36,6 +37,9 @@ public final class OkHttpConnectionFactory {
         return new OkHttpClient.Builder()
                 .cookieJar(SharedPreferenceCookieManager.getInstance())
                 .cache(NET_CACHE)
+                .connectTimeout(60, TimeUnit.SECONDS)
+                .writeTimeout(60, TimeUnit.SECONDS)
+                .readTimeout(60, TimeUnit.SECONDS)
                 .addInterceptor(getLoggingInterceptor())
                 .addInterceptor(new UnsuccessfulResponseInterceptor())
                 .addInterceptor(new CommonHeaderRequestInterceptor())

--- a/app/src/main/java/fr/free/nrw/commons/contributions/ChunkInfo.kt
+++ b/app/src/main/java/fr/free/nrw/commons/contributions/ChunkInfo.kt
@@ -6,20 +6,20 @@ import fr.free.nrw.commons.upload.UploadResult
 
 data class ChunkInfo(
     val uploadResult: UploadResult,
-    val lastChunkIndex: Int,
-    var isLastChunkUploaded: Boolean
+    val indexOfNextChunkToUpload: Int,
+    val totalChunks: Int
 ) : Parcelable {
     constructor(parcel: Parcel) : this(
         parcel.readParcelable(UploadResult::class.java.classLoader),
         parcel.readInt(),
-        parcel.readByte() != 0.toByte()
+        parcel.readInt()
     ) {
     }
 
     override fun writeToParcel(parcel: Parcel, flags: Int) {
         parcel.writeParcelable(uploadResult, flags)
-        parcel.writeInt(lastChunkIndex)
-        parcel.writeByte(if (isLastChunkUploaded) 1 else 0)
+        parcel.writeInt(indexOfNextChunkToUpload)
+        parcel.writeInt(totalChunks)
     }
 
     override fun describeContents(): Int {

--- a/app/src/main/java/fr/free/nrw/commons/db/AppDatabase.kt
+++ b/app/src/main/java/fr/free/nrw/commons/db/AppDatabase.kt
@@ -10,7 +10,7 @@ import fr.free.nrw.commons.contributions.ContributionDao
  * The database for accessing the respective DAOs
  *
  */
-@Database(entities = [Contribution::class], version = 5, exportSchema = false)
+@Database(entities = [Contribution::class], version = 6, exportSchema = false)
 @TypeConverters(Converters::class)
 abstract class AppDatabase : RoomDatabase() {
   abstract fun contributionDao(): ContributionDao

--- a/app/src/main/java/fr/free/nrw/commons/di/NetworkingModule.java
+++ b/app/src/main/java/fr/free/nrw/commons/di/NetworkingModule.java
@@ -62,7 +62,8 @@ public class NetworkingModule {
     public OkHttpClient provideOkHttpClient(Context context,
                                             HttpLoggingInterceptor httpLoggingInterceptor) {
         File dir = new File(context.getCacheDir(), "okHttpCache");
-        return new OkHttpClient.Builder().connectTimeout(60, TimeUnit.SECONDS)
+        return new OkHttpClient.Builder()
+            .connectTimeout(60, TimeUnit.SECONDS)
             .writeTimeout(60, TimeUnit.SECONDS)
                 .addInterceptor(httpLoggingInterceptor)
             .readTimeout(60, TimeUnit.SECONDS)

--- a/app/src/main/java/fr/free/nrw/commons/upload/FileUtilsWrapper.java
+++ b/app/src/main/java/fr/free/nrw/commons/upload/FileUtilsWrapper.java
@@ -45,7 +45,7 @@ public class FileUtilsWrapper {
   /**
    * Takes a file as input and returns an Observable of files with the specified chunk size
    */
-  public Observable<File> getFileChunks(Context context, File file, final int chunkSize)
+  public List<File> getFileChunks(Context context, File file, final int chunkSize)
       throws IOException {
     final byte[] buffer = new byte[chunkSize];
 
@@ -58,7 +58,7 @@ public class FileUtilsWrapper {
         buffers.add(writeToFile(context, Arrays.copyOf(buffer, size), file.getName(),
             getFileExt(file.getName())));
       }
-      return Observable.fromIterable(buffers);
+      return buffers;
     }
   }
 

--- a/app/src/main/java/fr/free/nrw/commons/upload/UploadClient.java
+++ b/app/src/main/java/fr/free/nrw/commons/upload/UploadClient.java
@@ -15,6 +15,7 @@ import io.reactivex.disposables.CompositeDisposable;
 import java.io.File;
 import java.io.IOException;
 import java.util.Date;
+import java.util.List;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicReference;
@@ -65,31 +66,39 @@ public class UploadClient {
   Observable<StashUploadResult> uploadFileToStash(
       final Context context, final String filename, final Contribution contribution,
       final NotificationUpdateProgressListener notificationUpdater) throws IOException {
-    if (contribution.getChunkInfo() != null && contribution.getChunkInfo().isLastChunkUploaded()) {
+    if (contribution.getChunkInfo() != null
+        && contribution.getChunkInfo().getTotalChunks() == contribution.getChunkInfo()
+        .getIndexOfNextChunkToUpload()) {
       return Observable.just(new StashUploadResult(StashUploadState.SUCCESS,
           contribution.getChunkInfo().getUploadResult().getFilekey()));
     }
     pauseUploads = false;
     File file = new File(contribution.getLocalUri().getPath());
-    final Observable<File> fileChunks = fileUtilsWrapper.getFileChunks(context, file, CHUNK_SIZE);
+    final List<File> fileChunks = fileUtilsWrapper.getFileChunks(context, file, CHUNK_SIZE);
+
+    int totalChunks = fileChunks.size();
+
     final MediaType mediaType = MediaType
         .parse(FileUtils.getMimeType(context, Uri.parse(file.getPath())));
 
-    final AtomicInteger index = new AtomicInteger();
+    final AtomicInteger indexOfNextChunkToBeUploaded = new AtomicInteger();
     final AtomicReference<ChunkInfo> chunkInfo = new AtomicReference<>();
     Timber.d("Chunk info");
     if (contribution.getChunkInfo() != null && isStashValid(contribution)) {
       chunkInfo.set(contribution.getChunkInfo());
+
+      for (int i = 0; i < contribution.getChunkInfo().getIndexOfNextChunkToUpload(); i++) {
+        Timber.d("Chunk: Removing chunks %d", i);
+        fileChunks.remove(i);
+      }
+
+      indexOfNextChunkToBeUploaded.set(contribution.getChunkInfo().getIndexOfNextChunkToUpload());
     }
 
     final AtomicBoolean failures = new AtomicBoolean();
 
-    compositeDisposable.add(fileChunks.forEach(chunkFile -> {
+    compositeDisposable.add(Observable.fromIterable(fileChunks).forEach(chunkFile -> {
       if (pauseUploads || failures.get()) {
-        return;
-      }
-      if (chunkInfo.get() != null && index.get() < chunkInfo.get().getLastChunkIndex()) {
-        index.getAndIncrement();
         return;
       }
       final int offset =
@@ -108,7 +117,10 @@ public class UploadClient {
           offset,
           filekey,
           countingRequestBody).subscribe(uploadResult -> {
-        chunkInfo.set(new ChunkInfo(uploadResult, index.incrementAndGet(), false));
+        Timber.d("Chunk: Uploaded chunk %d", indexOfNextChunkToBeUploaded.get());
+        Timber.d("Chunk: Queued chunk %d", indexOfNextChunkToBeUploaded.incrementAndGet());
+        chunkInfo.set(
+            new ChunkInfo(uploadResult, indexOfNextChunkToBeUploaded.get(), totalChunks));
         notificationUpdater.onChunkUploaded(contribution, chunkInfo.get());
       }, throwable -> {
         failures.set(true);
@@ -116,13 +128,12 @@ public class UploadClient {
       }));
     }));
 
-    chunkInfo.get().setLastChunkUploaded(true);
-    notificationUpdater.onChunkUploaded(contribution, chunkInfo.get());
     if (pauseUploads) {
       return Observable.just(new StashUploadResult(StashUploadState.PAUSED, null));
     } else if (failures.get()) {
       return Observable.just(new StashUploadResult(StashUploadState.FAILED, null));
     } else if (chunkInfo.get() != null) {
+      Timber.d("Going into stash upload success %s", chunkInfo.get());
       return Observable.just(new StashUploadResult(StashUploadState.SUCCESS,
           chunkInfo.get().getUploadResult().getFilekey()));
     } else {

--- a/app/src/main/java/fr/free/nrw/commons/upload/UploadClient.java
+++ b/app/src/main/java/fr/free/nrw/commons/upload/UploadClient.java
@@ -73,34 +73,35 @@ public class UploadClient {
           contribution.getChunkInfo().getUploadResult().getFilekey()));
     }
     pauseUploads = false;
-    File file = new File(contribution.getLocalUri().getPath());
+    final File file = new File(contribution.getLocalUri().getPath());
     final List<File> fileChunks = fileUtilsWrapper.getFileChunks(context, file, CHUNK_SIZE);
 
-    int totalChunks = fileChunks.size();
+    final int totalChunks = fileChunks.size();
 
     final MediaType mediaType = MediaType
         .parse(FileUtils.getMimeType(context, Uri.parse(file.getPath())));
 
     final AtomicInteger indexOfNextChunkToBeUploaded = new AtomicInteger();
     final AtomicReference<ChunkInfo> chunkInfo = new AtomicReference<>();
-    Timber.d("Chunk info");
-    if (contribution.getChunkInfo() != null && isStashValid(contribution)) {
+    if (isStashValid(contribution)) {
       chunkInfo.set(contribution.getChunkInfo());
-
-      for (int i = 0; i < contribution.getChunkInfo().getIndexOfNextChunkToUpload(); i++) {
-        Timber.d("Chunk: Removing chunks %d", i);
-        fileChunks.remove(i);
-      }
 
       indexOfNextChunkToBeUploaded.set(contribution.getChunkInfo().getIndexOfNextChunkToUpload());
     }
 
+    final AtomicInteger index = new AtomicInteger();
     final AtomicBoolean failures = new AtomicBoolean();
 
     compositeDisposable.add(Observable.fromIterable(fileChunks).forEach(chunkFile -> {
       if (pauseUploads || failures.get()) {
         return;
       }
+
+      if (chunkInfo.get() != null && index.get() < chunkInfo.get().getIndexOfNextChunkToUpload()) {
+        index.incrementAndGet();
+        return;
+      }
+      index.getAndIncrement();
       final int offset =
           chunkInfo.get() != null ? chunkInfo.get().getUploadResult().getOffset() : 0;
       final String filekey =
@@ -117,14 +118,11 @@ public class UploadClient {
           offset,
           filekey,
           countingRequestBody).subscribe(uploadResult -> {
-        Timber.d("Chunk: Uploaded chunk %d", indexOfNextChunkToBeUploaded.get());
-        Timber.d("Chunk: Queued chunk %d", indexOfNextChunkToBeUploaded.incrementAndGet());
         chunkInfo.set(
             new ChunkInfo(uploadResult, indexOfNextChunkToBeUploaded.get(), totalChunks));
         notificationUpdater.onChunkUploaded(contribution, chunkInfo.get());
       }, throwable -> {
         failures.set(true);
-        Timber.e(throwable, "Error occurred in uploading chunk");
       }));
     }));
 
@@ -133,7 +131,6 @@ public class UploadClient {
     } else if (failures.get()) {
       return Observable.just(new StashUploadResult(StashUploadState.FAILED, null));
     } else if (chunkInfo.get() != null) {
-      Timber.d("Going into stash upload success %s", chunkInfo.get());
       return Observable.just(new StashUploadResult(StashUploadState.SUCCESS,
           chunkInfo.get().getUploadResult().getFilekey()));
     } else {
@@ -147,8 +144,9 @@ public class UploadClient {
    * @return
    */
   private boolean isStashValid(Contribution contribution) {
-    return contribution.getDateModified()
-        .after(new Date(System.currentTimeMillis() - MAX_CHUNK_AGE));
+    return contribution.getChunkInfo() != null &&
+        contribution.getDateModified()
+            .after(new Date(System.currentTimeMillis() - MAX_CHUNK_AGE));
   }
 
   /**

--- a/app/src/main/java/fr/free/nrw/commons/upload/UploadClient.java
+++ b/app/src/main/java/fr/free/nrw/commons/upload/UploadClient.java
@@ -15,6 +15,7 @@ import io.reactivex.disposables.CompositeDisposable;
 import java.io.File;
 import java.io.IOException;
 import java.util.Date;
+import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicReference;
 import javax.inject.Inject;
@@ -80,8 +81,11 @@ public class UploadClient {
     if (contribution.getChunkInfo() != null && isStashValid(contribution)) {
       chunkInfo.set(contribution.getChunkInfo());
     }
+
+    final AtomicBoolean failures = new AtomicBoolean();
+
     compositeDisposable.add(fileChunks.forEach(chunkFile -> {
-      if (pauseUploads) {
+      if (pauseUploads || failures.get()) {
         return;
       }
       if (chunkInfo.get() != null && index.get() < chunkInfo.get().getLastChunkIndex()) {
@@ -107,6 +111,7 @@ public class UploadClient {
         chunkInfo.set(new ChunkInfo(uploadResult, index.incrementAndGet(), false));
         notificationUpdater.onChunkUploaded(contribution, chunkInfo.get());
       }, throwable -> {
+        failures.set(true);
         Timber.e(throwable, "Error occurred in uploading chunk");
       }));
     }));
@@ -115,6 +120,8 @@ public class UploadClient {
     notificationUpdater.onChunkUploaded(contribution, chunkInfo.get());
     if (pauseUploads) {
       return Observable.just(new StashUploadResult(StashUploadState.PAUSED, null));
+    } else if (failures.get()) {
+      return Observable.just(new StashUploadResult(StashUploadState.FAILED, null));
     } else if (chunkInfo.get() != null) {
       return Observable.just(new StashUploadResult(StashUploadState.SUCCESS,
           chunkInfo.get().getUploadResult().getFilekey()));

--- a/app/src/main/java/fr/free/nrw/commons/upload/UploadClient.java
+++ b/app/src/main/java/fr/free/nrw/commons/upload/UploadClient.java
@@ -15,7 +15,9 @@ import io.reactivex.disposables.CompositeDisposable;
 import java.io.File;
 import java.io.IOException;
 import java.util.Date;
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicReference;
@@ -43,7 +45,8 @@ public class UploadClient {
   private final PageContentsCreator pageContentsCreator;
   private final FileUtilsWrapper fileUtilsWrapper;
   private final Gson gson;
-  private boolean pauseUploads = false;
+
+  private Map<String, Boolean> pauseUploads;
 
   private final CompositeDisposable compositeDisposable = new CompositeDisposable();
 
@@ -57,6 +60,7 @@ public class UploadClient {
     this.pageContentsCreator = pageContentsCreator;
     this.fileUtilsWrapper = fileUtilsWrapper;
     this.gson = gson;
+    this.pauseUploads = new HashMap<>();
   }
 
   /**
@@ -72,7 +76,9 @@ public class UploadClient {
       return Observable.just(new StashUploadResult(StashUploadState.SUCCESS,
           contribution.getChunkInfo().getUploadResult().getFilekey()));
     }
-    pauseUploads = false;
+
+    pauseUploads.put(contribution.getPageId(), false);
+
     final File file = new File(contribution.getLocalUri().getPath());
     final List<File> fileChunks = fileUtilsWrapper.getFileChunks(context, file, CHUNK_SIZE);
 
@@ -81,29 +87,33 @@ public class UploadClient {
     final MediaType mediaType = MediaType
         .parse(FileUtils.getMimeType(context, Uri.parse(file.getPath())));
 
-    final AtomicInteger indexOfNextChunkToBeUploaded = new AtomicInteger();
     final AtomicReference<ChunkInfo> chunkInfo = new AtomicReference<>();
     if (isStashValid(contribution)) {
       chunkInfo.set(contribution.getChunkInfo());
 
-      indexOfNextChunkToBeUploaded.set(contribution.getChunkInfo().getIndexOfNextChunkToUpload());
+      Timber.d("Chunk: Next Chunk: %s, Total Chunks: %s",
+          contribution.getChunkInfo().getIndexOfNextChunkToUpload(),
+          contribution.getChunkInfo().getTotalChunks());
     }
 
     final AtomicInteger index = new AtomicInteger();
     final AtomicBoolean failures = new AtomicBoolean();
 
     compositeDisposable.add(Observable.fromIterable(fileChunks).forEach(chunkFile -> {
-      if (pauseUploads || failures.get()) {
+      if (pauseUploads.get(contribution.getPageId()) || failures.get()) {
         return;
       }
 
       if (chunkInfo.get() != null && index.get() < chunkInfo.get().getIndexOfNextChunkToUpload()) {
         index.incrementAndGet();
+        Timber.d("Chunk: Increment and return: %s", index.get());
         return;
       }
       index.getAndIncrement();
       final int offset =
           chunkInfo.get() != null ? chunkInfo.get().getUploadResult().getOffset() : 0;
+
+      Timber.d("Chunk: Sending Chunk number: %s, offset: %s", index.get(), offset);
       final String filekey =
           chunkInfo.get() != null ? chunkInfo.get().getUploadResult().getFilekey() : null;
 
@@ -118,15 +128,17 @@ public class UploadClient {
           offset,
           filekey,
           countingRequestBody).subscribe(uploadResult -> {
+        Timber.d("Chunk: Received Chunk number: %s, offset: %s", index.get(),
+            uploadResult.getOffset());
         chunkInfo.set(
-            new ChunkInfo(uploadResult, indexOfNextChunkToBeUploaded.get(), totalChunks));
+            new ChunkInfo(uploadResult, index.get(), totalChunks));
         notificationUpdater.onChunkUploaded(contribution, chunkInfo.get());
       }, throwable -> {
         failures.set(true);
       }));
     }));
 
-    if (pauseUploads) {
+    if (pauseUploads.get(contribution.getPageId())) {
       return Observable.just(new StashUploadResult(StashUploadState.PAUSED, null));
     } else if (failures.get()) {
       return Observable.just(new StashUploadResult(StashUploadState.FAILED, null));
@@ -182,9 +194,10 @@ public class UploadClient {
 
   /**
    * Dispose the active disposable and sets the pause variable
+   * @param pageId
    */
-  public void pauseUpload() {
-    pauseUploads = true;
+  public void pauseUpload(String pageId) {
+    pauseUploads.put(pageId, true);
     if (!compositeDisposable.isDisposed()) {
       compositeDisposable.dispose();
     }

--- a/app/src/main/java/fr/free/nrw/commons/upload/UploadService.java
+++ b/app/src/main/java/fr/free/nrw/commons/upload/UploadService.java
@@ -312,8 +312,6 @@ public class UploadService extends CommonsDaggerService {
         .flatMap(uploadStash -> {
           notificationManager.cancel(notificationTag, NOTIFICATION_UPLOAD_IN_PROGRESS);
 
-          Timber.d("Stash upload response 1 is %s", uploadStash.toString());
-
           if (uploadStash.getState() == StashUploadState.SUCCESS) {
             Timber.d("making sure of uniqueness of name: %s", filename);
             String uniqueFilename = findUniqueFilename(filename);
@@ -332,7 +330,6 @@ public class UploadService extends CommonsDaggerService {
               }
             });
           } else if (uploadStash.getState() == StashUploadState.PAUSED) {
-            Timber.d("Contribution upload paused");
             showPausedNotification(contribution);
             return Observable.never();
           } else {
@@ -359,7 +356,6 @@ public class UploadService extends CommonsDaggerService {
 
   private void onUpload(Contribution contribution, String notificationTag,
       UploadResult uploadResult) {
-    Timber.d("Stash upload response 2 is %s", uploadResult.toString());
 
     notificationManager.cancel(notificationTag, NOTIFICATION_UPLOAD_IN_PROGRESS);
 
@@ -401,7 +397,7 @@ public class UploadService extends CommonsDaggerService {
 
   @SuppressLint("StringFormatInvalid")
   @SuppressWarnings("deprecation")
-  private void showFailedNotification(Contribution contribution) {
+  private void showFailedNotification(final Contribution contribution) {
     final String displayTitle = contribution.getMedia().getDisplayTitle();
     curNotification.setTicker(getString(R.string.upload_failed_notification_title, displayTitle))
         .setContentTitle(getString(R.string.upload_failed_notification_title, displayTitle))
@@ -412,6 +408,7 @@ public class UploadService extends CommonsDaggerService {
         curNotification.build());
 
     contribution.setState(Contribution.STATE_FAILED);
+    contribution.setChunkInfo(null);
 
     compositeDisposable.add(contributionDao
         .update(contribution)
@@ -419,7 +416,7 @@ public class UploadService extends CommonsDaggerService {
         .subscribe());
   }
 
-  private void showPausedNotification(Contribution contribution) {
+  private void showPausedNotification(final Contribution contribution) {
     final String displayTitle = contribution.getMedia().getDisplayTitle();
     curNotification.setTicker(getString(R.string.upload_paused_notification_title, displayTitle))
         .setContentTitle(getString(R.string.upload_paused_notification_title, displayTitle))

--- a/app/src/main/java/fr/free/nrw/commons/upload/UploadService.java
+++ b/app/src/main/java/fr/free/nrw/commons/upload/UploadService.java
@@ -148,7 +148,7 @@ public class UploadService extends CommonsDaggerService {
    * @param contribution
    */
   public void pauseUpload(Contribution contribution) {
-    uploadClient.pauseUpload();
+    uploadClient.pauseUpload(contribution.getPageId());
     contribution.setState(Contribution.STATE_PAUSED);
     compositeDisposable.add(contributionDao.update(contribution)
         .subscribeOn(ioThreadScheduler)

--- a/build.gradle
+++ b/build.gradle
@@ -7,7 +7,7 @@ buildscript {
         maven { url "https://plugins.gradle.org/m2/" }
     }
     dependencies {
-        classpath 'com.android.tools.build:gradle:4.0.0'
+        classpath 'com.android.tools.build:gradle:4.0.1'
         classpath "com.hiya:jacoco-android:0.2"
         classpath 'com.getkeepsafe.dexcount:dexcount-gradle-plugin:0.8.2'
         classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$KOTLIN_VERSION"

